monitoring/grafana/dashboards/rbd-details.json

```diff
@@ -0,0 +1,412 @@
+      "thresholds": [],
+      "timeFrom": null,
+      "timeShift": null,
+      "title": "IO Bytes per Second",
+      "tooltip": {
+        "shared": true,
+        "sort": 0,
+        "value_type": "individual"
+      },
+      "type": "graph",
+      "xaxis": {
+        "buckets": null,
+        "mode": "time",
+        "name": null,
+        "show": true,
+        "values": []
+      },
+      "yaxes": [
+        {
+          "format": "Bps",
+          "label": null,
+          "logBase": 1,
+          "max": null,
+          "min": "0",
+          "show": true
+        },
+        {
+          "format": "Bps",
+          "label": null,
+          "logBase": 1,
+          "max": null,
+          "min": null,
+          "show": true
+        }
+      ],
+      "yaxis": {
+        "align": true,
+        "alignLevel": null
+      }
+    },
+    {
+      "aliasColors": {},
+      "bars": false,
+      "dashLength": 10,
+      "dashes": false,
+      "datasource": "$Datasource",
+      "fill": 1,
+      "gridPos": {
+        "h": 9,
+        "w": 8,
+        "x": 16,
+        "y": 0
+      },
+      "id": 2,
+      "legend": {
+        "avg": false,
+        "current": false,
+        "max": false,
+        "min": false,
+        "show": true,
+        "total": false,
+        "values": false
+      },
+      "lines": true,
+      "linewidth": 1,
+      "links": [],
+      "nullPointMode": "null as zero",
+      "percentage": false,
+      "pointradius": 5,
+      "points": false,
+      "renderer": "flot",
+      "seriesOverrides": [],
+      "spaceLength": 10,
+      "stack": false,
+      "steppedLine": false,
+      "targets": [
+        {
+          "expr": "round(sum(irate(ceph_rbd_write_latency_sum{pool=\"$Pool\", image=\"$Image\"}[30s])) / sum(irate(ceph_rbd_write_latency_count{pool=\"$Pool\", image=\"$Image\"}[30s])))",
+          "format": "time_series",
+          "intervalFactor": 1,
+          "legendFormat": "Write Latency Sum",
+          "refId": "A"
+        },
+        {
+          "expr": "round(sum(irate(ceph_rbd_read_latency_sum{pool=\"$Pool\", image=\"$Image\"}[30s])) / sum(irate(ceph_rbd_read_latency_count{pool=\"$Pool\", image=\"$Image\"}[30s])))",
+          "format": "time_series",
+          "intervalFactor": 1,
+          "legendFormat": "Read Latency Sum",
+          "refId": "B"
+        }
+      ],
+      "thresholds": [],
+      "timeFrom": null,
+      "timeShift": null,
+      "title": "Averange Latency",
+      "tooltip": {
+        "shared": true,
+        "sort": 0,
+        "value_type": "individual"
+      },
+      "type": "graph",
+      "xaxis": {
+        "buckets": null,
+        "mode": "time",
+        "name": null,
+        "show": true,
+        "values": []
+      },
+      "yaxes": [
+        {
+          "format": "ns",
+          "label": null,
+          "logBase": 1,
+          "max": null,
+          "min": "0",
+          "show": true
+        },
+        {
+          "format": "ns",

```

Comment:
Are you sure prometheus exports nano seconds here? Looking at the screenshot seems to agree that there is a bit of a magnitude issue here.

Reply:
If that would be wrong in my dashboard, then this would also be wrong in the summary dashboard. In there `ns` is also applied.

Reply:
The metrics itself are exported in miliseconds accoridng to the help. The irate function calculates a per second rate. So we have `round(sum(irate(ms))/sum(irate(ms))) = round(sum(s)/sum(s)) = round(ns)`? Which doesn't make sense for me at all. But if I increase to miliseconds then the values because not really realistic. The values in my screenshot could be correct, since I wrote a lot of data to the disks at these times. If you think this is correct please resolve this discussion.

Reply:
> The metrics itself are exported in miliseconds accoridng to the help. The irate function calculates a per second rate. So we have `round(sum(irate(ms))/sum(irate(ms))) = round(sum(s)/sum(s)) = round(ns)`? 
As you mentioned yourself we deal with ops per second (irate) so your unit derivation doesn't quite work that way. Though I'm actually not 100% sure that formula is correct to begin with, do we really want to divide a rate by another rate?

Reply:
Well I don't know. I just took what was on the overview page and then adjusted it for the details. The formula is comming from there. I am certainly no ceph expert...

Reply:
> I just took what was on the overview page and then adjusted it for the details. The formula is comming from there. I am certainly no ceph expert...

Good thing this has nothing to do with ceph :)

Lets look at the query. We have a latency sum and latency count, i.e. we don't get a latency metric, we need to calculate one. Using `sum/count` gives us the average latency, no need for any irate or sum calls.

If we want the current latency though `irate(ceph_rbd_read_latency_sum{pool=\"$Pool\", image=\"$Image\"}[30s]) (for the read side) should be enough though.

Reply:
And we should change the queries in other dashboards as well. The way they are now makes no sense to me.

Reply:
@jan--f Thanks for the input. I will change that in my dashboards, as well as in the other dashboards.

Reply:
Alright, took another look. My earlier comments were not quite accurate either.
So what is this plot about. Lets go by the panel title (the legend text claims something different): the average latencies.

In this case the formular `irate(ceph_rbd_write_latency_sum) / irate(ceph_rbd_write_latency_count)` works but the units are off. We have `(ms/s)/(ops/s)` which resolves to `ms/op`. What purpose the sum and round serve is not obvious to me.

Another mistake I made earlier was this:

> If we want the current latency though irate(ceph_rbd_read_latency_sum{pool=\"$Pool\", image=\"$Image\"}[30s]) (for the read side) should be enough though.

This only gives us the rate of the sum of course, which is useless since this has no info on how many ops there were in the same timeframe.